firmware/src/core/utils/HTTPClientWrapper.cpp

```diff
@@ -0,0 +1,167 @@
+#include "HTTPClientWrapper.h"
+
+HTTPClientWrapper* HTTPClientWrapper::instance = nullptr;
+SemaphoreHandle_t HTTPClientWrapper::requestSemaphore = nullptr;
+QueueHandle_t HTTPClientWrapper::requestQueue = nullptr;
+QueueHandle_t HTTPClientWrapper::responseQueue = nullptr;
+
+// Initialize static debug counters
+volatile uint32_t HTTPClientWrapper::activeRequests = 0;
+volatile uint32_t HTTPClientWrapper::maxConcurrentRequests = 0;
+
+HTTPClientWrapper::HTTPClientWrapper() {
+    if (!requestSemaphore) {
+        requestSemaphore = xSemaphoreCreateBinary();
+        xSemaphoreGive(requestSemaphore);
+    }
+    if (!requestQueue) {
+        requestQueue = xQueueCreate(REQUEST_QUEUE_SIZE, REQUEST_QUEUE_ITEM_SIZE);  
+    }
+    if (!responseQueue) {
+        responseQueue = xQueueCreate(RESPONSE_QUEUE_SIZE, RESPONSE_QUEUE_ITEM_SIZE);
+    }
+}
+
+HTTPClientWrapper::~HTTPClientWrapper() {
+    if (requestSemaphore) {
+        vSemaphoreDelete(requestSemaphore);
+        requestSemaphore = nullptr;
+    }
+    if (requestQueue) {
+        vQueueDelete(requestQueue);
+        requestQueue = nullptr;
+    }
+    if (responseQueue) {
+        vQueueDelete(responseQueue);
+        responseQueue = nullptr;
+    }
+}
+
+HTTPClientWrapper* HTTPClientWrapper::getInstance() {
+    if (!instance) {
+        instance = new HTTPClientWrapper();
+    }
+    return instance;
+}
+
+bool HTTPClientWrapper::addRequest(const String& url, ResponseCallback callback) {
+    auto* params = new RequestParams{url, callback};
+    
+    if (xQueueSend(requestQueue, &params, 0) != pdPASS) {
+        delete params;
+        Serial.println("Failed to queue HTTP request");
+        return false;
+    }
+    
+    Serial.printf("Queued request: %s (Queue length: %d)\n", url.c_str(), uxQueueMessagesWaiting(requestQueue));
+    return true;
+}
+
+void HTTPClientWrapper::processRequestQueue() {  
+    // First check if there are any requests to process
+    if (uxQueueMessagesWaiting(requestQueue) == 0) {
+        return; // No requests in queue
+    }
+
+    // Only try to take semaphore if we have work to do
+    if (xSemaphoreTake(requestSemaphore, 0) != pdTRUE) {
+        //Serial.println("⚠️ Semaphore blocked - request already in progress");
+        return;
+    }
+
+    Serial.println("✅ Obtained semaphore");
+    activeRequests++;
```

Comment:
What is the point of this?
The semaphore will guard it, so this is always one (when busy) or 0 (when not busy) ?

Reply:
correct.  it was just a double check while debugging to ensure that only one at max